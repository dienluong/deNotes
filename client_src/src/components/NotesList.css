/* NotesList styles */
.active-tree-node {
  box-shadow: 0 0 0 4px pink;
}
