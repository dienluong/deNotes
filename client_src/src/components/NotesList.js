/* TODO: Revisit this component and react-ui-tree code to see if rendering can be optimized
   https://reactjs.org/docs/optimizing-performance.html
*/
// import 'react-ui-tree/dist/react-ui-tree.css';
// import Tree from 'react-ui-tree';
// TODO: Delete Note.js
// import Note from './Note';

import Tree, { removeNode } from 'react-sortable-tree';
import 'react-sortable-tree/style.css';
// TODO: Delete NotesList.css
// import './NotesList.css';
import React from 'react';

import sampleNotes from '../test/sample-tree';

class NotesList extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      notesTree: sampleNotes,
    };

    this.handleChange = this.handleChange.bind(this);
    // this.renderNode = this.renderNode.bind(this);
    // this.onClickNode = this.onClickNode.bind(this);
    this.deleteNote = this.deleteNote.bind(this);
  }

  handleChange(notesTree) {
    this.setState({
      notesTree,
    });
  }

  /* TODO: Remove, formerly for react-ui-tree
  renderNode(node) {
    // console.log(`renderNode ${node.module}`);
    return (
      <Note
        active={ this.state.activeNote === node }
        clickHandler={ this.onClickNode }
        node={ node }
      />
    );
  }

  onClickNode(node) {
    // console.log(`Active Note: ${node.module}`);
    this.setState({
      activeNote: node,
    });
  }
  */

  deleteNote({ path }) {
    return ({
      buttons: [
        <button
          style={{ verticalAlign: 'middle' }}
          onClick={() => {
            const { treeData } = removeNode({
              treeData: this.state.notesTree,
              getNodeKey: ({ treeIndex }) => treeIndex,
              path,
            });
            this.setState({
              notesTree: treeData,
            });
          }}
        >
          x
        </button>,
      ],
    });
  }

  render() {
    console.log('rendering NoteLists........................');
    return (
      <Tree
        className='tree'
        treeData={ this.state.notesTree }
        onChange={ this.handleChange }
        generateNodeProps={ this.deleteNote }
      />
    );
  };
}

export default NotesList;
