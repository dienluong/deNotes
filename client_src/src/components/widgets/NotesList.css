/* NotesList styles */
.dnt__active-tree-node {
  box-shadow: 0 0 0 4px pink;
}

.dnt__tree-node-btn {
  vertical-align: middle;
}
