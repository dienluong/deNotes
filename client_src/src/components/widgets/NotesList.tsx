import React from 'react';
import Toolbar from './Toolbar';
import PathNavigator from './PathNavigator';
import Tree from 'react-sortable-tree';
import mobileTheme from '../../tree-theme';
import 'react-sortable-tree/style.css';
import styles from './NotesList.module.css';
import NodeTitle from './NodeTitle';

// Types
import { TreeItem } from 'react-sortable-tree';
type PropsT = {
  tree: TreeNodeT[],
  activeNode: ActiveNodeT,
  activePath: string[],
  treeChangeHandler: (...args: any) => any;
  nodeTitleChangeHandler: (...args: any) => any,
  nodeClickHandler: (params: { id: TreeNodeT["id"], path: TreeNodePathT }) => unknown,
  nodeDoubleClickHandler: (params: { id: TreeNodeT["id"], path: TreeNodePathT }) => unknown,
  deleteNodeBtnHandler: (params: { node: TreeNodeT, path: TreeNodePathT }) => unknown,
  addNoteBtnHandler: (params: { path: TreeNodePathT }) => unknown,
  pathNavigatorClickHandler: (...args: any) => any,
  toolbarHandlersMap: any,
  getNodeKey: (...args: any) => any,
};
type generateNodePropsT = ({ node, path }: { node: TreeItem, path: (string|number)[] }) => object;

function NotesList({
  tree,
  activeNode,
  activePath,
  treeChangeHandler,
  nodeTitleChangeHandler,
  nodeClickHandler,
  nodeDoubleClickHandler,
  deleteNodeBtnHandler,
  addNoteBtnHandler,
  pathNavigatorClickHandler,
  toolbarHandlersMap,
  getNodeKey,
}: PropsT) {
  function buildNodeProps({ node, path }: { node: TreeNodeT, path: string[] }) {
    return ({
      title: (<NodeTitle node={ node } path={ path } onSubmit={ nodeTitleChangeHandler } />),
      className: (node.id === activeNode.id) ? `${styles['dnt__tree-node']} ${styles['dnt__tree-node--active']}` : styles['dnt__tree-node'],
      buttons: _buildNodeButtons({ node, path }),
      tabIndex: '0',
      'data-testid': node.id,
      onClick: () => nodeClickHandler({ id: node.id, path }),
      onDoubleClick: () => nodeDoubleClickHandler({ id: node.id, path }),
    });
  }

  function _buildNodeButtons({ node, path }: { node: TreeNodeT, path: string[] }) {
    let buttons = [
      <button
        className={ styles['dnt__tree-node-btn'] }
        onClick={ (event) => {
          event.stopPropagation();
          deleteNodeBtnHandler({ node, path });
        }}
      >
        x
      </button>,
    ];

    // Check if current node is parent node; '+' button only for parent nodes
    if (typeof node.children !== 'undefined') {
      buttons.unshift(
        <button
          className={ styles['dnt__tree-node-btn'] }
          onClick={ (event) => {
            event.stopPropagation();
            addNoteBtnHandler({ path });
          }}
        >
          +
        </button>);
    }

    return buttons;
  }

  function _collapseAllParents({ tree }: { tree: TreeNodeT[] }): TreeNodeT[] {
    return tree.map((node): TreeNodeT => {
      if (node && node.children) {
        return { ...node, expanded: false };
      } else {
        return node;
      }
    });
  }

  return (
    <div className={ styles["dnt__notes-list"] }>
      <Toolbar toolsMap={ toolbarHandlersMap } />
      <PathNavigator
        path={ activePath }
        activeSegmentIdx={ activeNode.path.indexOf(activeNode.id) }
        onClick={ pathNavigatorClickHandler }
      />
      <Tree
        className={ 'tree ' + styles.dnt__tree }
        treeData={ _collapseAllParents({ tree }) }
        theme={ mobileTheme as any }
        onChange={ treeChangeHandler }
        getNodeKey={ getNodeKey }
        generateNodeProps={ buildNodeProps as generateNodePropsT }
      />
    </div>
  );
}

export default NotesList;
