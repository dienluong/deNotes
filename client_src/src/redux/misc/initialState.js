import uuid from 'uuid/v1';
import Delta from 'quill-delta';
const _ID_DELIMITER = process.env.REACT_APP_ID_DELIMITER;

// TODO: Is it the best place to define this?
const _rootNode = [
  {
    title: '/',
    subtitle: '',
    uniqid: uuid(),
    get id() {
      return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
    },
    type: 'folder',
    expanded: true,
    children: [],
  },
];

const epoch = 0;
export default {
  notesTree: _rootNode,
  activeNode: {
    id: _rootNode[0].id,
    path: [_rootNode[0].id],
  },
  editorContent: {
    id: '',
    title: '',
    content: '<p><br></p>',
    delta: new Delta(),
    dateCreated: epoch,
    dateModified: epoch,
    readOnly: false,
  },
};
