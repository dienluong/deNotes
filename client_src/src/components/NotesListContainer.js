import { connect } from 'react-redux';
import NotesList from './widgets/NotesList';
import {
  selectNodeAction,
  navigatePathAction,
  changeNotesTreeAction,
  changeNodeTitleAction,
  deleteNodeAction,
  addNoteAction,
  switchActiveNodeOnDelete,
  switchActiveNodeOnAdd,
  addAndSelectNode,
} from '../redux/actions/notesListActions';

import { getNodeKey, translatePathToInfo } from '../utils/treeUtils';

function mapStateToProps(state) {
  const activePath = translatePathToInfo({ notesTree: state.notesTree, path: state.activeNode.path, kind: 'title' });
  return {
    notesTree: state.notesTree,
    activeNode: state.activeNode,
    activePath,
  };
}

function mapDispatchToProps(dispatch) {
  function toolbarNewFolderBtnClickHandler() {
    return dispatch(addAndSelectNode({ kind: 'folder' }));
  }

  function toolbarNewNoteBtnClickHandler() {
    return dispatch(addAndSelectNode({ kind: 'item' }));
  }

  const toolbarHandlersMap = new Map();

  toolbarHandlersMap.set('New Folder', toolbarNewFolderBtnClickHandler);
  toolbarHandlersMap.set('New Note', toolbarNewNoteBtnClickHandler);

  return {
    treeChangeHandler(notesTree) {
      return dispatch(changeNotesTreeAction(notesTree));
    },
    nodeTitleChangeHandler({ node, title, path }) {
      return dispatch(changeNodeTitleAction({ node, title, path }));
    },
    pathNavigatorClickHandler({ idx }) {
      return dispatch(navigatePathAction({ idx }));
    },
    nodeClickHandler({ id = null, path = [] }) {
      return dispatch(selectNodeAction({ id, path }));
    },
    deleteNodeBtnHandler({ node, path }) {
      dispatch(deleteNodeAction({ node, path }));
      return dispatch(switchActiveNodeOnDelete({ id: node.id, path }));
    },
    addNoteBtnHandler({ path }) {
      dispatch(addNoteAction({ path }));
      return dispatch(switchActiveNodeOnAdd({ path }));
    },
    toolbarHandlersMap: toolbarHandlersMap,
  };
}

function mergeProps(stateProps, dispatchProps, ownProps) {
  return Object.assign({}, ownProps, stateProps, dispatchProps, { getNodeKey });
}

const NotesListContainer = connect(mapStateToProps, mapDispatchToProps, mergeProps)(NotesList);
export default NotesListContainer;
