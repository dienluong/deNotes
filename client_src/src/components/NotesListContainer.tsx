import { connect } from 'react-redux';
import { selectTitlesFromActivePath, selectSiblingsOfActiveNode } from '../redux/selectors';
import * as rootReducer from '../redux/reducers';
import NotesList from './widgets/NotesList';
import {
  selectNodeThunkAction,
  navigatePathThunkAction,
  changeNotesFolderThunkAction,
  changeNodeTitleAction,
  deleteNodeThunkAction,
  addAndSelectNodeThunkAction,
} from '../redux/actions/notesListActions';

import { getNodeKey } from '../utils/treeUtils';

// Types
import { ThunkDispatch } from 'redux-thunk';
import { AnyAction } from 'redux';
interface DispatchProps {
  treeChangeHandler: (params: any) => AnyAction;
  nodeTitleChangeHandler: (params: { node: TreeNodeT, title: string, path: TreeNodePathT }) => AnyAction;
  pathNavigatorClickHandler: (params: { idx: number }) => AnyAction;
  nodeClickHandler: (params: any) => AnyAction;
  deleteNodeBtnHandler: (params: any) => Promise<AnyAction>;
  addNoteBtnHandler: (params: any) => AnyAction;
  toolbarHandlersMap: Map<string, () => AnyAction>;
}

export const TOOLBAR_LABELS = {
  NEW_FOLDER: 'New Folder',
  NEW_NOTE: 'New Note',
};

function mapStateToProps(state: AppStateT) {
  const activePathByTitles = selectTitlesFromActivePath(state);

  // const activePath = translatePathToInfo({ notesTree: state.notesTree, path: state.activeNode.path, kind: 'title' });
  if (mapStateToProps.cache.notesTree !== rootReducer.selectNotesTree(state)) {
    console.log('~~~~~~~~~~~~~~~~ Notes Tree Changed');
    mapStateToProps.cache.notesTree = rootReducer.selectNotesTree(state);
  } else {
    console.log('~~~~~~~~~~~~~~~~ Same Notes Tree');
  }
  if (mapStateToProps.cache.activeNode !== rootReducer.selectActiveNode(state)) {
    console.log('~~~~~~~~~~~~~~~~ Active Node Changed');
    mapStateToProps.cache.activeNode = rootReducer.selectActiveNode(state);
  } else {
    console.log('~~~~~~~~~~~~~~~~ Same Active Node');
  }
  if (mapStateToProps.cache.activePath !== activePathByTitles) {
    console.log('~~~~~~~~~~~~~~~~ Active Path Changed');
    mapStateToProps.cache.activePath = activePathByTitles;
  } else {
    console.log('~~~~~~~~~~~~~~~~ Same Active Path');
  }
  if (mapStateToProps.cache.editorContent !== rootReducer.selectEditorContent(state)) {
    console.log('~~~~~~~~~~~~~~~~ Editor Content Changed');
    mapStateToProps.cache.editorContent = rootReducer.selectEditorContent(state);
  } else {
    console.log('~~~~~~~~~~~~~~~~ Same Editor Content');
  }
  // TODO: Remove ABOVE

  // Get the siblings of the current active node (including itself). The children will be passed as prop to the component
  const children = selectSiblingsOfActiveNode(state);
  return {
    tree: children,
    activeNode: rootReducer.selectActiveNode(state),
    activePath: activePathByTitles,
  };
}

// TODO: Remove
// Memoization.
mapStateToProps.cache = {} as { notesTree: any, activeNode: any, activePath: any, editorContent: any };

function mapDispatchToProps(dispatch: ThunkDispatch<AppStateT, any, AnyAction>): DispatchProps {
  function toolbarNewFolderBtnHandler() {
    return dispatch(addAndSelectNodeThunkAction({ kind: 'folder' }));
  }

  function toolbarNewNoteBtnHandler() {
    return dispatch(addAndSelectNodeThunkAction({ kind: 'item' }));
  }

  const toolbarHandlersMap = new Map();

  toolbarHandlersMap.set(TOOLBAR_LABELS.NEW_FOLDER, toolbarNewFolderBtnHandler);
  toolbarHandlersMap.set(TOOLBAR_LABELS.NEW_NOTE, toolbarNewNoteBtnHandler);

  return {
    treeChangeHandler(tree) {
      return dispatch(changeNotesFolderThunkAction({ folder: tree }));
    },
    nodeTitleChangeHandler({ node, title, path }) {
      // We are not using the path received from the NotesList component because that path is not for the entire tree;
      // remember that NotesList only receives and renders a given leaf of the whole tree.
      return dispatch(changeNodeTitleAction({ node, title }));
    },
    pathNavigatorClickHandler({ idx }) {
      return dispatch(navigatePathThunkAction({ idx }));
    },
    nodeClickHandler({ id = '', path = [] }) {
      // We are not using the path received from the NotesList component because that path is not for the entire tree;
      // remember that NotesList only receives and renders a given leaf of the whole tree.
      return dispatch(selectNodeThunkAction({ id }));
    },
    deleteNodeBtnHandler({ node, path }) {
      // We are not using the path received from the NotesList component because that path is not for the entire tree;
      // remember that NotesList only receives and renders a given leaf of the whole tree.
      return dispatch(deleteNodeThunkAction({ node }));
    },
    addNoteBtnHandler({ path } ) {
      // We are not using the path received from the NotesList component because that path is not for the entire tree;
      // remember that NotesList only receives and renders a given leaf of the whole tree.
      return dispatch(addAndSelectNodeThunkAction({ kind: 'item' }));
    },
    toolbarHandlersMap: toolbarHandlersMap,
  };
}

function mergeProps(stateProps: object, dispatchProps: object, ownProps: object) {
  return Object.assign({}, ownProps, stateProps, dispatchProps, { getNodeKey });
}

const NotesListContainer = connect(mapStateToProps, mapDispatchToProps, mergeProps)(NotesList);
export default NotesListContainer;
