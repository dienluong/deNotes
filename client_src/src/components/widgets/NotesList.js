import React, { Fragment } from 'react';
import Toolbar from './Toolbar';
import PathNavigator from './PathNavigator';
import Tree from 'react-sortable-tree';
import 'react-sortable-tree/style.css';
import './NotesList.css';
import NodeTitle from './NodeTitle';

function NotesList({
  notesTree,
  activeNode,
  activePath,
  treeChangeHandler,
  nodeTitleChangeHandler,
  nodeClickHandler,
  deleteNodeBtnHandler,
  addNoteBtnHandler,
  pathNavigatorClickHandler,
  toolbarHandlersMap,
  getNodeKey,
}) {
  function buildNodeProps({ node, path }) {
    return ({
      title: (<NodeTitle node={ node } path={ path } onSubmit={ nodeTitleChangeHandler } />),
      className: (node.id === activeNode.id) ? 'dnt__active-tree-node' : '',
      buttons: _buildNodeButtons({ node, path }),
      tabIndex: '0',
      onClick: () => nodeClickHandler({ id: node.id, path }),
      onFocus: () => console.log('Focused: ' + node.title),
      onBlur: () => console.log('unFocused: ' + node.title),
    });
  }

  function _buildNodeButtons({ node, path }) {
    let buttons = [
      <button
        className='dnt__tree-node-btn'
        onClick={ (event) => {
          event.stopPropagation();
          deleteNodeBtnHandler({ node, path });
        }}
      >
        x
      </button>,
    ];

    // Check if current node is parent node; '+' button only for parent nodes
    if (typeof node.children !== 'undefined') {
      buttons.unshift(
        <button
          className='dnt__tree-node-btn'
          onClick={ (event) => {
            event.stopPropagation();
            addNoteBtnHandler({ path });
          }}
        >
          +
        </button>);
    }

    return buttons;
  }

  // TODO: remove
  // console.log(`
  //   Tree: ${JSON.stringify(notesTree)} \n
  //   Active ID: ${activeNode.id} \n
  //   Path: ${activeNode.path} \n
  //   ${ activePath }
  // `);

  return (
    <Fragment>
      <Toolbar toolsMap={ toolbarHandlersMap } />
      <PathNavigator
        path={ activePath }
        activeSegmentIdx={ activeNode.path.indexOf(activeNode.id) }
        onClick={ pathNavigatorClickHandler }
      />
      <Tree
        className='tree'
        treeData={ notesTree }
        onChange={ treeChangeHandler }
        getNodeKey={ getNodeKey }
        generateNodeProps={ buildNodeProps }
      />
    </Fragment>
  );
}

export default NotesList;
