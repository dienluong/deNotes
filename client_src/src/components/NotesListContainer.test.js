import React from 'react';
import createMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { render, cleanup, fireEvent } from 'react-testing-library';
import 'jest-dom/extend-expect';
import NotesListContainer, { TOOLBAR_LABELS } from './NotesListContainer';
import NotesList from './widgets/NotesList';
import initialState from '../redux/misc/initialState';
import { mockedTree } from '../test-utils/mocks/mockedNotesTree';
import { getNodeKey, collapseFolders } from '../utils/treeUtils';
jest.mock('../redux/actions/notesListActions');
import {
  selectNodeThunkAction,
  navigatePathThunkAction,
  goUpAFolderAction,
  changeNotesFolderThunkAction,
  changeNodeTitleAction,
  deleteNodeThunkAction,
  addAndSelectNodeThunkAction,
} from '../redux/actions/notesListActions';

import pathNavStyles from './widgets/PathNavigator.module.css';
import notesListStyles from './widgets/NotesList.module.css';
import nodeTitleStyles from './widgets/NodeTitle.module.css';

// Configure Enzyme
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
let wrapper = {};

afterEach(() => {
  if (typeof wrapper.unmount === 'function') {
    wrapper.unmount();
  }
  cleanup();
  addAndSelectNodeThunkAction.mockClear();
  navigatePathThunkAction.mockClear();
  goUpAFolderAction.mockClear();
  selectNodeThunkAction.mockClear();
  changeNodeTitleAction.mockClear();
  deleteNodeThunkAction.mockClear();
  changeNotesFolderThunkAction.mockClear();
});

it('renders correctly with prop values taken from redux store', () => {
  const mockedNotesTree = {
    id: '1234',
    tree: mockedTree,
    dateCreated: 1111,
    dateModified: 2222,
  };
  const mockedInitialState = {
    ...initialState,
    notesTree: mockedNotesTree,
    activeNode: {
      id: mockedNotesTree.tree[0].id,
      path: [mockedNotesTree.tree[0].id],
    },
  };

  // ---> test case where active folder is not root: all sub-folders should be rendered collapsed
  const expectedProps = {
    tree: collapseFolders({ tree: mockedNotesTree.tree[0].children }),
    activeNode: { ...mockedInitialState.activeNode },
    activePath: [mockedNotesTree.tree[0].title],
    getNodeKey,
  };
  const store = createMockStore([thunk])(mockedInitialState);
  const { container } = render(<Provider
    store={store}><NotesListContainer/></Provider>);
  expect(container).toMatchSnapshot();

  wrapper = shallow(<Provider
    store={store}><NotesListContainer/></Provider>).dive({ context: { store } });

  expect(wrapper.find(NotesList).exists()).toBe(true);
  expect(wrapper.find(NotesList).props()).toMatchObject(expectedProps);

});

it('renders root folder with all its sub-folders are expanded ', () => {
  const mockedNotesTree = {
    id: '1234',
    tree: mockedTree,
    dateCreated: 1111,
    dateModified: 2222,
  };
  const mockedInitialState = {
    ...initialState,
    notesTree: mockedNotesTree,
    activeNode: {
      id: mockedNotesTree.tree[0].id,
      path: [mockedNotesTree.tree[0].id],
    },
  };

  // ---> test case where active folder is root: all its sub-folders should be rendered expanded;
  // In the case below, the active node is an ITEM node (as opposed to a FOLDER) of the root, so the active folder is root itself.
  mockedInitialState.activeNode = {
    id: mockedNotesTree.tree[1].id,
    path: [mockedNotesTree.tree[1].id],
  };

  const expectedProps = {
    tree: mockedNotesTree.tree,
    activeNode: { ...mockedInitialState.activeNode },
    activePath: [mockedNotesTree.tree[1].title],
    getNodeKey,
  };

  const store = createMockStore([thunk])(mockedInitialState);
  const { container } = render(<Provider store={ store }><NotesListContainer /></Provider>);
  expect(container).toMatchSnapshot();

  wrapper = shallow(<Provider store={ store }><NotesListContainer /></Provider>).dive({ context: { store } });

  expect(wrapper.find(NotesList).exists()).toBe(true);
  expect(wrapper.find(NotesList).props()).toMatchObject(expectedProps);
});

it('reacts to events by calling proper handlers', () => {
  const mockedNotesTree = { id: '1234', tree: mockedTree, dateCreated: 1111, dateModified: 2222 };
  const mockedInitialState = {
    ...initialState,
    notesTree: mockedNotesTree,
    activeNode: {
      id: mockedNotesTree.tree[0].id,
      path: [mockedNotesTree.tree[0].id],
    },
  };
  addAndSelectNodeThunkAction.mockImplementation(() => ({ type: 'DUMMY_ACTION', payload: 'DUMMY_PAYLOAD' }));
  navigatePathThunkAction.mockImplementation(() => ({ type: 'DUMMY_ACTION', payload: 'DUMMY_PAYLOAD' }));
  goUpAFolderAction.mockImplementation(() => ({ type: 'DUMMY_ACTION', payload: 'DUMMY_PAYLOAD' }));
  selectNodeThunkAction.mockImplementation(() => ({ type: 'DUMMY_ACTION', payload: 'DUMMY_PAYLOAD' }));
  changeNodeTitleAction.mockImplementation(() => ({ type: 'DUMMY_ACTION', payload: 'DUMMY_PAYLOAD' }));
  deleteNodeThunkAction.mockImplementation(() => ({ type: 'DUMMY_ACTION', payload: 'DUMMY_PAYLOAD' }));

  const store = createMockStore([thunk])(mockedInitialState);
  // const store = createStore((state) => state, mockedInitialState, applyMiddleware(thunk));

  const { queryAllByText, container } = render(<Provider store={ store }><NotesListContainer /></Provider>);

  // Test toolbar clicks
  let target = queryAllByText(TOOLBAR_LABELS.NEW_FOLDER);
  // wrapper.find(`.${toolStyles.dnt__tool}[children="New Folder"]`).simulate('click');
  expect(target).toHaveLength(1);
  fireEvent.click(target[0]);
  expect(addAndSelectNodeThunkAction).toBeCalledTimes(1);

  target = queryAllByText(TOOLBAR_LABELS.NEW_NOTE);
  expect(target).toHaveLength(1);
  fireEvent.click(target[0]);
  expect(addAndSelectNodeThunkAction).toBeCalledTimes(2);
  addAndSelectNodeThunkAction.mockClear();

  target = queryAllByText(TOOLBAR_LABELS.BACK_BUTTON);
  expect(target).toHaveLength(1);
  fireEvent.click(target[0]);
  expect(goUpAFolderAction).toBeCalledTimes(1);
  goUpAFolderAction.mockClear();

  // Test navbar click
  target = container.querySelectorAll(`.${pathNavStyles["dnt__pathnav-segment--active"]}`);
  expect(target).toHaveLength(1);
  fireEvent.click(target[0]);
  expect(navigatePathThunkAction).toBeCalledTimes(1);

  // Test node click
  target = container.querySelectorAll(`.${notesListStyles["dnt__tree-node"]}`);
  // our choosen mocked tree has more than one node
  expect(target.length).toBeGreaterThan(1);
  // arbitrarily click on second node
  fireEvent.click(target[1]);
  expect(selectNodeThunkAction).toBeCalledTimes(1);

  // Test title change on blur
  target = container.querySelectorAll(`.${nodeTitleStyles["dnt__node-title"]}`);
  expect(target.length).toBeGreaterThan(1);
  // change the title of the second node
  target[1].value = 'changed title';
  fireEvent.blur(target[1]);
  expect(changeNodeTitleAction).toBeCalledTimes(1);

  // Test node button '+' click
  target = queryAllByText('+');
  // our mocked tree has more than one node
  expect(target.length).toBeGreaterThan(1);
  // just arbitrarily choose to click on '+' button of the second node
  fireEvent.click(target[1]);
  expect(addAndSelectNodeThunkAction).toBeCalledTimes(1);

  // Test node button 'x' click
  target = queryAllByText('x');
  // our mocked tree has more than one node
  expect(target.length).toBeGreaterThan(1);
  // just arbitrarily choose to click on '+' button of the second node
  fireEvent.click(target[1]);
  expect(deleteNodeThunkAction).toBeCalledTimes(1);
});
