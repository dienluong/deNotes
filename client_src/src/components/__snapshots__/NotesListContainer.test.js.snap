// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with prop values taken from redux store 1`] = `
<div>
  <div
    class="dnt__notes-list"
  >
    <div
      class="dnt__toolbar"
    >
      <button
        class="dnt__tool"
      >
        New Folder
      </button>
      <button
        class="dnt__tool"
      >
        New Note
      </button>
    </div>
    <nav
      class="dnt__pathnav"
    >
      <ul>
        <li
          class="lined dnt__pathnav-segment--active"
        >
          <span>
            folder root0
            &gt;
          </span>
        </li>
      </ul>
    </nav>
    <div
      class="rst__tree tree dnt__tree"
      style="height: 100%;"
    >
      <div>
        <div
          aria-label="grid"
          aria-readonly="true"
          class="ReactVirtualized__Grid ReactVirtualized__List rst__virtualScrollOverride"
          role="grid"
          style="box-sizing: border-box; direction: ltr; height: 99999px; position: relative; width: 200px; overflow-x: hidden; overflow-y: hidden;"
          tabindex="0"
        >
          <div
            class="ReactVirtualized__Grid__innerScrollContainer"
            role="rowgroup"
            style="width: auto; height: 248px; max-width: 200px; max-height: 248px; overflow: hidden; position: relative;"
          >
            <div
              class="node"
              style="height: 62px; left: 0px; position: absolute; top: 0px; width: 100%;"
            >
              <div
                class="nodeContent"
                style="padding-left: 0px;"
              >
                <div
                  data-testid="ITEM_NODE|^|2"
                  style="height: 100%;"
                  tabindex="0"
                >
                  <div
                    class="rowWrapper"
                  >
                    <div
                      class="row dnt__tree-node"
                      style="opacity: 1; padding-left: 45px;"
                    >
                      <div
                        class="rowContents"
                        draggable="true"
                      >
                        <div
                          class="rowLabel"
                        >
                          <span
                            class="rowTitle"
                          >
                            <form>
                              <input
                                class="dnt__node-title"
                                type="text"
                                value="note root0.children0"
                              />
                            </form>
                          </span>
                          
                        </div>
                        <div
                          class="rowToolbar"
                        >
                          <div
                            class="toolbarButton"
                          >
                            <button
                              class="dnt__tree-node-btn"
                            >
                              x
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="node"
              style="height: 62px; left: 0px; position: absolute; top: 62px; width: 100%;"
            >
              <div
                class="nodeContent"
                style="padding-left: 0px;"
              >
                <div
                  data-testid="ITEM_NODE|^|3"
                  style="height: 100%;"
                  tabindex="0"
                >
                  <div
                    class="rowWrapper"
                  >
                    <div
                      class="row dnt__tree-node"
                      style="opacity: 1; padding-left: 45px;"
                    >
                      <div
                        class="rowContents"
                        draggable="true"
                      >
                        <div
                          class="rowLabel"
                        >
                          <span
                            class="rowTitle"
                          >
                            <form>
                              <input
                                class="dnt__node-title"
                                type="text"
                                value="note root0.children1"
                              />
                            </form>
                          </span>
                          
                        </div>
                        <div
                          class="rowToolbar"
                        >
                          <div
                            class="toolbarButton"
                          >
                            <button
                              class="dnt__tree-node-btn"
                            >
                              x
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="node"
              style="height: 62px; left: 0px; position: absolute; top: 124px; width: 100%;"
            >
              <div
                class="nodeContent"
                style="padding-left: 0px;"
              >
                <div
                  data-testid="FOLDER_NODE|^|4"
                  style="height: 100%;"
                  tabindex="0"
                >
                  <div>
                    <button
                      aria-label="Collapse"
                      class="expandButton"
                      type="button"
                    />
                  </div>
                  <div
                    class="rowWrapper"
                  >
                    <div
                      class="row dnt__tree-node"
                      style="opacity: 1; padding-left: 45px;"
                    >
                      <div
                        class="rowContents"
                        draggable="true"
                      >
                        <div
                          class="rowLabel"
                        >
                          <span
                            class="rowTitle"
                          >
                            <form>
                              <input
                                class="dnt__node-title"
                                type="text"
                                value="folder root0.children2"
                              />
                            </form>
                          </span>
                          
                        </div>
                        <div
                          class="rowToolbar"
                        >
                          <div
                            class="toolbarButton"
                          >
                            <button
                              class="dnt__tree-node-btn"
                            >
                              +
                            </button>
                          </div>
                          <div
                            class="toolbarButton"
                          >
                            <button
                              class="dnt__tree-node-btn"
                            >
                              x
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="node"
              style="height: 62px; left: 0px; position: absolute; top: 186px; width: 100%;"
            >
              <div
                class="nodeContent"
                style="padding-left: 0px;"
              >
                <div
                  data-testid="FOLDER_NODE|^|7"
                  style="height: 100%;"
                  tabindex="0"
                >
                  <div
                    class="rowWrapper"
                  >
                    <div
                      class="row dnt__tree-node"
                      style="opacity: 1; padding-left: 45px;"
                    >
                      <div
                        class="rowContents"
                        draggable="true"
                      >
                        <div
                          class="rowLabel"
                        >
                          <span
                            class="rowTitle"
                          >
                            <form>
                              <input
                                class="dnt__node-title"
                                type="text"
                                value="folder root0.children3"
                              />
                            </form>
                          </span>
                          
                        </div>
                        <div
                          class="rowToolbar"
                        >
                          <div
                            class="toolbarButton"
                          >
                            <button
                              class="dnt__tree-node-btn"
                            >
                              +
                            </button>
                          </div>
                          <div
                            class="toolbarButton"
                          >
                            <button
                              class="dnt__tree-node-btn"
                            >
                              x
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
